google/cloud/spanner/client.cc

```diff
@@ -99,16 +99,74 @@ StatusOr<std::int64_t> Client::ExecutePartitionedDml(
   return Status(StatusCode::kUnimplemented, "not implemented");
 }
 
-StatusOr<CommitResult> Client::Commit(
-    Transaction const& /*transaction*/,
-    std::vector<Mutation> const& /*mutations*/) {
-  return Status(StatusCode::kUnimplemented, "not implemented");
+StatusOr<CommitResult> Client::Commit(Transaction transaction,
+                                      std::vector<Mutation> const& mutations) {
+  using F = std::function<StatusOr<CommitResult>(
+      spanner_proto::TransactionSelector&)>;
+  return internal::Visit(
+      std::move(transaction),
+      F([this, &mutations](spanner_proto::TransactionSelector& s) {
+        return this->Commit(s, mutations);
+      }));
 }
 
 Status Client::Rollback(Transaction const& /*transaction*/) {
   return Status(StatusCode::kUnimplemented, "not implemented");
 }
 
+Client::SessionHolder::SessionHolder(std::string session,
+                                     Client* client) noexcept
+    : session_(std::move(session)), client_(client) {}
+
+Client::SessionHolder::~SessionHolder() {
+  client_->sessions_.emplace_back(std::move(session_));
+}
+
+StatusOr<Client::SessionHolder> Client::GetSession() {
+  if (!sessions_.empty()) {
+    std::string session = sessions_.back();
+    sessions_.pop_back();
+    return SessionHolder(std::move(session), this);
+  }
+  grpc::ClientContext context;
+  google::spanner::v1::CreateSessionRequest request;
+  request.set_database(database_name_);
+  auto response = stub_->CreateSession(context, request);
+  if (!response) {
+    return response.status();
+  }
+  return SessionHolder(std::move(*response->mutable_name()), this);
+}
+
+StatusOr<CommitResult> Client::Commit(
+    google::spanner::v1::TransactionSelector& s,
+    std::vector<Mutation> const& mutations) {
+  auto session = GetSession();
+  if (!session) {
+    return std::move(session).status();
+  }
+  google::spanner::v1::CommitRequest request;
+  request.set_session(session->session_name());
+  for (auto const& m : mutations) {
+    *request.add_mutations() = m.as_proto();
+  }
+  if (!s.id().empty()) {
+    request.set_transaction_id(s.id());
+  } else if (s.has_begin()) {
+    *request.mutable_single_use_transaction() = s.begin();
```

Comment:
I'm not sure why `CommitRequest` has a special `oneof transaction` instead of using `TransactionSelector` like everything else, apart from driving home the fact that "begin" really is "single-use" for commits.. So, I agree that this is the right thing to do.  For any actual read-write transaction (i.e., not just write) we will have already bound an id.

Reply:
Ack. I take it that no changes are required.